tensorflow/core/profiler/internal/gpu/rocm_tracer.h

```diff
@@ -0,0 +1,302 @@
+  uint32 thread_id = kInvalidThreadId;
+  int64 stream_id = kInvalidStreamId;
+  union {
+    MemcpyDetails memcpy_info;      // If type == Memcpy*
+    MemsetDetails memset_info;      // If type == Memset*
+    MemAllocDetails memalloc_info;  // If type == MemoryAlloc
+    KernelDetails kernel_info;      // If type == Kernel
+  };
+};
+
+void DumpRocmTracerEvent(const RocmTracerEvent& event, uint64 start_walltime_ns,
+                         uint64 start_gputime_ns);
+
+struct RocmTracerOptions {
+  // map of domain --> ops for which we need to enable the API callbacks
+  // If the ops vector is empty, then enable API callbacks for entire domain
+  std::map<activity_domain_t, std::vector<uint32_t> > api_callbacks;
+
+  // map of domain --> ops for which we need to enable the Activity records
+  // If the ops vector is empty, then enable Activity records for entire domain
+  std::map<activity_domain_t, std::vector<uint32_t> > activity_tracing;
+};
+
+struct RocmTraceCollectorOptions {
+  // Maximum number of events to collect from callback API; if -1, no limit.
+  // if 0, the callback API is enabled to build a correlation map, but no
+  // events are collected.
+  uint64 max_callback_api_events;
+  // Maximum number of events to collect from activity API; if -1, no limit.
+  uint64 max_activity_api_events;
+  // Maximum number of annotation strings that we can accommodate.
+  uint64 max_annotation_strings;
+  // Number of GPUs involved.
+  uint32 num_gpus;
+};
+
+class AnnotationMap {
+ public:
+  explicit AnnotationMap(uint64 max_size) : max_size_(max_size) {}
+  void Add(uint32 correlation_id, const std::string& annotation);
+  absl::string_view LookUp(uint32 correlation_id);
+
+ private:
+  struct AnnotationMapImpl {
+    // The population/consumption of annotations might happen from multiple
+    // callback/activity api related threads.
+    absl::Mutex mutex;
+    // Annotation tends to be repetitive, use a hash_set to store the strings,
+    // an use the reference to the string in the map.
+    absl::node_hash_set<std::string> annotations;
+    absl::flat_hash_map<uint32, absl::string_view> correlation_map;
+  };
+  const uint64 max_size_;
+  AnnotationMapImpl map_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(AnnotationMap);
+};
+
+class RocmTraceCollector {
+ public:
+  explicit RocmTraceCollector(const RocmTraceCollectorOptions& options)
+      : options_(options), annotation_map_(options.max_annotation_strings) {}
+  virtual ~RocmTraceCollector() {}
+
+  virtual void AddEvent(RocmTracerEvent&& event) = 0;
+  virtual void OnEventsDropped(const std::string& reason,
+                               uint32 num_events) = 0;
+  virtual void Flush() = 0;
+
+  AnnotationMap* annotation_map() { return &annotation_map_; }
+
+ protected:
+  RocmTraceCollectorOptions options_;
+
+ private:
+  AnnotationMap annotation_map_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(RocmTraceCollector);
+};
+
+// forward declarations for callback functors
+class RocmApiCallbackImpl;
```

Comment:
Our style guide discourages forward declaration and prefers to include headers directly (include what you use).

Reply:
those class implementations are in the `rocm_tracer.cc` file, so no option but to forward declare them.

Reply:
What is commonly done is to declare the class interface in the header, and leave the implementation in the .cc file.

Reply:
> What is commonly done is to declare the class interface in the header, and leave the implementation in the .cc file.

True, but that also adds unnecessary code to the `rocm_tracer.h` file and affects both readability and maintainability. In this case, the two classes that are forward declared, are accesses only within `rocm_tracer.cc` and adding their interface to the `.h` file provides no additional value. I do not see any valid reason for pulling their interface into the header.

>  Our style guide discourages forward declaration and prefers to include headers directly (include what you use).

agree with you on the style guideline, but I am hoping I have made a case here for an exception.